openssh: Add legacy_dsa variant.

fhgwright opened the pull request:
Also adds message regarding test requirements.

Closes: https://trac.macports.org/ticket/70319

TESTED:
With +legacy_dsa, accepts "+ssh-rsa,ssh-dss" config options. Built successfully on OSX 10.4-10.5 ppc, 10.4-10.6 i386, 10.5-12.x x86_64, and 11.x-14.x arm64.  Included all variants compatible with available dependencies on the respective platforms, except that testing of +universal was limited to avoid the need to install +universal versions of many dependencies.

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
```
Mac OS X 10.4.11 8S165, PPC, Xcode 2.5 8M2558
Mac OS X 10.4.11 8S2167, i386, Xcode 2.5 8M2558
Mac OS X 10.5.8 9L31a, PPC, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, i386, Xcode 3.1.4 9M2809
Mac OS X 10.5.8 9L31a, x86_64, Xcode 3.1.4 9M2809
Mac OS X 10.6.8 10K549, i386, Xcode 3.2.6 10M2518
Mac OS X 10.6.8 10K549, x86_64, Xcode 3.2.6 10M2518
Mac OS X 10.7.5 11G63, x86_64, Xcode 4.6.3 4H1503
OS X 10.8.5 12F2560, x86_64, Xcode 5.1.1 5B1008
OS X 10.9.5 13F1911, x86_64, Xcode 6.2 6C131e
OS X 10.10.5 14F2511, x86_64, Xcode 7.2 7C68
OS X 10.11.6 15G22010, x86_64, Xcode 8.1 8B62
macOS 10.12.6 16G2136, x86_64, Xcode 9.2 9C40b
macOS 10.13.6 17G14042, x86_64, Xcode 10.1 10B61
macOS 10.14.6 18G9323, x86_64, Xcode 11.3.1 11C505
macOS 10.15.7 19H15, x86_64, Xcode 12.4 12D4e
macOS 11.7.10 20G1427, x86_64, Xcode 13.2.1 13C100
macOS 11.7.10 20G1427, arm64, Xcode 13.2.1 13C100
macOS 12.7.6 21H1320, x86_64, Xcode 14.2 14C18
macOS 12.7.6 21H1320, arm64, Xcode 14.2 14C18
macOS 13.6.8 22G820, arm64, Xcode 15.2 15C500b
macOS 14.6 23G80, arm64, Xcode 15.4 15F31d
```

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [NO] tried existing tests with `sudo port test`? `*`
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

`*` - Tests require an appropriate cooperating server.

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

artkiver commented:
@fhgwright thanks for this!

Builds OK for me on 14.6.1 as well.

I did notice this: 

```
port lint --nitpick
--->  Verifying Portfile for openssh
Warning: Unnecessary platforms line as darwin is the default
--->  0 errors and 1 warnings found.

```

Which, I don't recall the last time I ran port lint --nitpick on the Portfile before a PR, but I think my version of MacPorts has recently been updated since then so maybe port lint is being a bit more persnickety than it was previously as well? Regardless, that is kind of minor, just something I noticed in my testing locally that I thought I would mention.

Thank you for testing on such a wide variety of older OS versions!

neverpanic commented:
Should we add a paragraph to the note text that users should migrate their keys *now* and not show up shouting when this variant eventually breaks?

artkiver commented:
> Should we add a paragraph to the note text that users should migrate their keys _now_ and not show up shouting when this variant eventually breaks?

That seems as if it might be a good idea!

I'm not too sure about the wording.

Maybe something like:

"For users still utilizing DSA, support will be removed from OpenSSH entirely in 2025. It is recommended to generate and distribute new keys to relevant hosts using a non-deprecated cipher, e.g. 'sshkeygen -t ed25519'."

?

pmetzger commented on net/openssh/Portfile:
```diff
@@ -162,6 +162,11 @@ if {${name} eq ${subport}} {
+        ui_msg "Tests require a cooperating server named 'somehost'."
+        ui_msg "Failure to connect to it results in a hang."
+    }
+
```
Why does it require a remote host? One could just connect to an unprivileged server spun up and listening on an unusual port on localhost, no?

fhgwright replied:
@pmetzger 
> Why does it require a remote host? One could just connect to an unprivileged server spun up and listening on an unusual port on localhost, no?

Don't shoot the messenger. :-)

This message is simply reporting the upstream brain damage, in hopes of saving others the trouble of letting `port test` sit there for a long time, thinking that it's a really time-consuming test, when in reality it's just a failure to meet the unspecified prerequisites, in combination with a "hang on failure" test design.

And yes, the server really does need to be named 'somehost', presumably by adding an entry to `/etc/hosts`.  It's not even a well-localized reference:
```
MacPro:openssh-portable fw$ git grep -w somehost | wc -l
     237
```

Most references are to just 'somehost', but there are also a few of these:
```
MacPro:openssh-portable fw$ git grep -w somehost.example.com | wc -l
       3
```

fhgwright commented:
@artkiver 
> @fhgwright thanks for this!
> 
> Builds OK for me on 14.6.1 as well.
> 
> I did notice this:
> 
> ```
> port lint --nitpick
> --->  Verifying Portfile for openssh
> Warning: Unnecessary platforms line as darwin is the default
> --->  0 errors and 1 warnings found.
> ```
> 
> Which, I don't recall the last time I ran port lint --nitpick on the Portfile before a PR, but I think my version of MacPorts has recently been updated since then so maybe port lint is being a bit more persnickety than it was previously as well? Regardless, that is kind of minor, just something I noticed in my testing locally that I thought I would mention.

I believe that warning was just added in 2.10.0, and I'd run `port lint --nitpick` before upgrading base.

@neverpanic 
> Should we add a paragraph to the note text that users should migrate their keys _now_ and not show up shouting when this variant eventually breaks?

I guess there could be some more explanation, though realistically, `port notes` doesn't have sufficient visibility for this sort of issue.  A warning on each use would be better, but that would require a code patch, rather than just something in the Portfile.  And to do it really well, it would need to be a sufficiently elaborate patch to distinguish between the cases where the option is provided and not needed versus being provided and needed.

And in the case where DSA is required for the host key on a system that the user has no control over, the only recourse will be to disable host key checking for that system completely.

artkiver reviewed (approved):
Seems consistent with my experiences with port test kind of hanging; whereas if I run make test from the OpenSSH source directly from the tarball, it will definitely take a while to run but complete successfully.

artkiver commented:
> @artkiver
> 
> > @fhgwright thanks for this!
> > Builds OK for me on 14.6.1 as well.
> > I did notice this:
> > ```
> > port lint --nitpick
> > --->  Verifying Portfile for openssh
> > Warning: Unnecessary platforms line as darwin is the default
> > --->  0 errors and 1 warnings found.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which, I don't recall the last time I ran port lint --nitpick on the Portfile before a PR, but I think my version of MacPorts has recently been updated since then so maybe port lint is being a bit more persnickety than it was previously as well? Regardless, that is kind of minor, just something I noticed in my testing locally that I thought I would mention.
> 
> I believe that warning was just added in 2.10.0, and I'd run `port lint --nitpick` before upgrading base.

Thanks for the perspective/sanity check on that observation!

> 
> @neverpanic
> 
> > Should we add a paragraph to the note text that users should migrate their keys _now_ and not show up shouting when this variant eventually breaks?
> 
> I guess there could be some more explanation, though realistically, `port notes` doesn't have sufficient visibility for this sort of issue. A warning on each use would be better, but that would require a code patch, rather than just something in the Portfile. And to do it really well, it would need to be a sufficiently elaborate patch to distinguish between the cases where the option is provided and not needed versus being provided and needed.
> 
> And in the case where DSA is required for the host key on a system that the user has no control over, the only recourse will be to disable host key checking for that system completely.

Modifying the code to warn on each use seems as if it would be a much bigger ordeal, even if a more verbose port notes isn't really the greatest alternative, hopefully this isn't a variant that is going to be used widely and those who use it are at least made aware that they're outliers bucking the trend?

fhgwright commented:
@artkiver 
> Seems consistent with my experiences with port test kind of hanging; whereas if I run make test from the OpenSSH source directly from the tarball, it will definitely take a while to run but complete successfully.

I admit that I didn't try that.  I set up a repo with the intent of fixing the test phase if it was easy, but once I came across the stuff trying to connect to 'somehost', I figured that it would take more effort than I wanted to spend, so I just added the message.

It's not clear that enabling the test phase at all makes sense under the circumstances, but I suppose if someone sufficiently familiar with the requirements sets up the appropriate environment, it might actually work.

It's a good thing the CI tests don't run `port test`. :-)

neverpanic commented:
> I guess there could be some more explanation, though realistically, `port notes` doesn't have sufficient visibility for this sort of issue. A warning on each use would be better, but that would require a code patch, rather than just something in the Portfile. And to do it really well, it would need to be a sufficiently elaborate patch to distinguish between the cases where the option is provided and not needed versus being provided and needed.

Such a code patch should be done upstream. The problem of visibility isn't limited to MacPorts.


> And in the case where DSA is required for the host key on a system that the user has no control over, the only recourse will be to disable host key checking for that system completely.

And yet, it would allow users to contact the responsible admins for that host.

artkiver commented:
Just checking in here, what else should we be doing on our end?

I am guessing upstream isn't going to be too open to warning with DSA since they were making release announcements about its eventual deprecation going back to 2015? If someone wants to open a BugZilla issue with OpenSSH around this, they're more than welcome to if they think that is the way to move forward, but I am not really going to spearhead such an effort personally.

The variant as it stands, seems mostly OK from my vantage.

fhgwright commented:
@neverpanic 
> > I guess there could be some more explanation, though realistically, `port notes` doesn't have sufficient visibility for this sort of issue. A warning on each use would be better, but that would require a code patch, rather than just something in the Portfile. And to do it really well, it would need to be a sufficiently elaborate patch to distinguish between the cases where the option is provided and not needed versus being provided and needed.
> 
> Such a code patch should be done upstream. The problem of visibility isn't limited to MacPorts.

True, but it's not clear that they'd be amenable to such a patch.

> > And in the case where DSA is required for the host key on a system that the user has no control over, the only recourse will be to disable host key checking for that system completely.
> 
> And yet, it would allow users to contact the responsible admins for that host.

With variable success.

@artkiver 
> Just checking in here, what else should we be doing on our end?

Yeah, I wasn't aware that expanding the note was considered a blocking requirement.

> I am guessing upstream isn't going to be too open to warning with DSA since they were making release announcements about its eventual deprecation going back to 2015? If someone wants to open a BugZilla issue with OpenSSH around this, they're more than welcome to if they think that is the way to move forward, but I am not really going to spearhead such an effort personally.

Indeed.  They seem to assume that all users religiously read the release notes on every release, and implement their own reminders about when action (sometimes not at all trivial) is needed.

> The variant as it stands, seems mostly OK from my vantage.

I added a couple of lines to the notes, and fixed the "platforms" warning.

Force-pushed twice, to separate the rebase from the real change.

neverpanic commented:
> @neverpanic
> 
> > > I guess there could be some more explanation, though realistically, `port notes` doesn't have sufficient visibility for this sort of issue. A warning on each use would be better, but that would require a code patch, rather than just something in the Portfile. And to do it really well, it would need to be a sufficiently elaborate patch to distinguish between the cases where the option is provided and not needed versus being provided and needed.
> > 
> > Such a code patch should be done upstream. The problem of visibility isn't limited to MacPorts.
> 
> True, but it's not clear that they'd be amenable to such a patch.

Have you asked or tried?
Unless they reject such a patch, I really don't think MacPorts should add something like this downstream, and even if they reject it, I'd argue MacPorts isn't the place to add this warning.

> > And yet, it would allow users to contact the responsible admins for that host.
> 
> With variable success.

None of which would be MacPorts' fault. MacPorts cannot fix every problem with every open source software or its users, and neither should it want to. Upstream is the place where these things should be improved.

neverpanic commented:
> Just checking in here, what else should we be doing on our end?
> […]
> The variant as it stands, seems mostly OK from my vantage.

Nothing, just didn't notice that you were the maintainer and had already approved.
